File: problems/0001-two-sum/solution.erl
-spec two_sum(Nums :: [integer()], Target :: integer()) -> [integer()].
two_sum(Nums, Target) ->
    two_sum_(Nums, Target, #{}, 0).

two_sum_([], _Target, _Map, _Idx) ->
    {error, not_found};
two_sum_([N | R], Target, Map, Idx) ->
    Complement = Target - N,
    case maps:get(Complement, Map, undefined) of
        undefined ->
            two_sum_(R, Target, Map#{N => Idx}, Idx + 1);
        CompIdx ->
            [CompIdx, Idx]
    end.

File: problems/0001-two-sum/solution.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    Enum.reduce_while(nums, {%{}, 0}, fn num, {map, idx} ->
      complement = target - num

      case map[complement] do
        nil -> {:cont, {Map.put(map, num, idx), idx + 1}}
        comp_idx -> {:halt, [comp_idx, idx]}
      end
    end)
  end
end
